### ISSUE TITLE
Shared library not found when running tests (libtest)

### ISSUE THREAD

**Issue by redshirtrob** [bazelbuild/rules_rust#183]
# Overview

When running a `rust_test` that depends on `quote` and `syn` I get the following error:
```
/home/rob/.cache/bazel/_bazel_rob/72eb038da5e5b7f4c56e935524a9a14b/sandbox/linux-sandbox/1/execroot/__main__/bazel-out/k8-fastbuild/bin/lib/test_proc.runfiles/__main__/lib/test_proc: error while loading shared libraries: libtest-c0479f7cd36f68a3.so: cannot open shared object file: No such file or directory
```
The same test passes using `cargo test`.

## Show Me The Code
Since this is a pretty weird issue, I've generated a minimal project that exhibits the issue:

https://github.com/redshirtrob/rust-bazel-proc_macro-test

## Details
 - Rust 1.31.0
 - Bazel Version 0.19.2
 - proc-macro2 0.4.9
 - quote 0.6.4
 - syn 0.15.22

I've verified the library in question exists within the Bazel sandbox.

**Comment by mfarrugi** [comment]
tl;dr this is almost definitely an upstream bug.

---

(Your rules_rust and cargo-raze versions are a bit behind, but they are not the cause of this problem.)

The binary built by cargo test fails the same way when run directly,
```
>> /home/marco/.cache/cargo/debug/deps/test_proc-36289756094210ee
/home/marco/.cache/cargo/debug/deps/test_proc-36289756094210ee: error while loading shared libraries: libtest-c0479f7cd36f68a3.so: cannot open shared object file: No such file or directory
```
so this means `bazel test` is not providing something that `cargo test` does.

`cargo test --verbose --verbose` doesn't show the environment (https://github.com/rust-lang/cargo/pull/6017) used to run the binary, but providing LD_LIBRARY_PATH is a possible fix:
```
LD_LIBRARY_PATH=./bazel-rust-bazel-proc_macro-test/external/rust_linux_x86_64/lib ./bazel-bin/lib/test_proc

running 1 test
test test_fail ... ok

test result: ok. 1 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

Those libraries are dynamically linked, which rules_rust does not expect. I am not familiar with syn/quote et al, and so I don't know why they are being dynamically linked at all. AFAIU these are proc_macros being compiled into the test binary? If this is valid `rust_test` will need to look more like `rust_doc_test` where we create a script that provides `LD_LIBRARY_PATH`, or we can add the rpaths to the binary.

This is almost definitely an upstream bug, alluded to in https://github.com/rust-lang/cargo/issues/4651#issuecomment-357536550 and https://github.com/djc/askama/issues/73 .. perhaps fixed by https://github.com/djc/askama/pull/77

---

Things I looked at to poke around:

Bazel with rules_rust on master, and latest cargo-raze:
```
>> bazel build -s //lib:test_proc

INFO: Analysed target //lib:test_proc (0 packages loaded).
INFO: Found 1 target...
SUBCOMMAND: # //lib:test_proc [action 'Compiling Rust lib test_proc (1 files)']
(cd /home/marco/.cache/bazel/_bazel_marco/cad6d7f67a3f7a5033b35c378d6fa130/execroot/__main__ && \
  exec env - \
    CARGO_PKG_AUTHORS='' \
    CARGO_PKG_DESCRIPTION='' \
    CARGO_PKG_HOMEPAGE='' \
    CARGO_PKG_NAME=test_proc \
    CARGO_PKG_VERSION=0.0.0 \
    CARGO_PKG_VERSION_MAJOR=0 \
    CARGO_PKG_VERSION_MINOR=0 \
    CARGO_PKG_VERSION_PATCH=0 \
    CARGO_PKG_VERSION_PRE='' \
  /nix/store/x3nxn1vbxbpn005h1p82ivmwbjc4n94a-bash/bin/bash -c
        'CARGO_MANIFEST_DIR=$(pwd)/lib external/rust_linux_x86_64/bin/rustc "$@" --remap-path-prefix="$(pwd)"=__bazel_redacted_pwd'
        ''
        lib/tests/test_proc.rs
        '--crate-name=test_proc'
        '--crate-type=lib'
        '--codegen=metadata='
        '--out-dir=bazel-out/k8-fastbuild/bin/lib'
        '--codegen=extra-filename='
        '--codegen=opt-level=0'
        '--codegen=debuginfo=0'
        '--emit=dep-info,link'
        '--color=always'
        '--target=x86_64-unknown-linux-gnu' --test
        '--codegen=linker=/usr/bin/gcc'
        --codegen 'link-args=-m64 -fuse-ld=gold -Wl,-no-as-needed -Wl,-z,relro,-z,now -B/usr/bin -B/usr/bin -pass-exit-codes -lstdc++ -lm -m64'
        --extern 'quote=bazel-out/k8-fastbuild/bin/external/raze__quote__0_6_4/libquote--1216550202.rlib'
        --extern 'syn=bazel-out/k8-fastbuild/bin/external/raze__syn__0_15_22/libsyn--1006831486.rlib'
        --extern 'proc=bazel-out/k8-fastbuild/bin/lib/libproc--916610401.rlib'
        '-Ldependency=bazel-out/k8-fastbuild/bin/external/raze__quote__0_6_4'
        '-Ldependency=bazel-out/k8-fastbuild/bin/external/raze__proc_macro2__0_4_9'
        '-Ldependency=bazel-out/k8-fastbuild/bin/external/raze__unicode_xid__0_1_0'
        '-Ldependency=bazel-out/k8-fastbuild/bin/external/raze__syn__0_15_22'
        '-Ldependency=bazel-out/k8-fastbuild/bin/lib'
)
Target //lib:test_proc up-to-date:
  bazel-bin/lib/test_proc
INFO: Elapsed time: 0.441s, Critical Path: 0.38s
INFO: 1 process: 1 linux-sandbox.
INFO: Build completed successfully, 2 total actions
```

compared with cargo test
```
>> cargo test --verbose --verbose
    ...
rustc
    --crate-name test_proc tests/test_proc.rs
    --color always
    --emit=dep-info,link
    -C debuginfo=2
    --test
    --cfg 'feature="default"'
    -C metadata=36289756094210ee
    -C extra-filename=-36289756094210ee
    --out-dir /home/marco/.cache/cargo/debug/deps
    -C incremental=/home/marco/.cache/cargo/debug/incremental
    -L dependency=/home/marco/.cache/cargo/debug/deps
    --extern proc=/home/marco/.cache/cargo/debug/deps/libproc-92fc007963201f11.rlib
    --extern proc_macro2=/home/marco/.cache/cargo/debug/deps/libproc_macro2-f564b8a685496207.rlib
    --extern quote=/home/marco/.cache/cargo/debug/deps/libquote-e2eff3b567ce22ab.rlib
    --extern syn=/home/marco/.cache/cargo/debug/deps/libsyn-d9211a917353f357.rlib
```

Libraries are dynamically linked..
```
>> ldd /home/marco/.cache/cargo/debug/deps/test_proc-36289756094210ee
         linux-vdso.so.1 (0x00007ffc4875e000)
         libtest-c0479f7cd36f68a3.so => not found
         libproc_macro-520699affe05435a.so => not found
         libsyntax-7c956b82cf4d08b4.so => not found
         libsyntax_pos-aab3d379f3226375.so => not found
         libstd-52f862a21e09568c.so => not found
         libgcc_s.so.1 => /lib/x86_64-linux-gnu/libgcc_s.so.1 (0x00007f024047d000)
         libc.so.6 => /lib/x86_64-linux-gnu/libc.so.6 (0x00007f024008c000)
         /lib64/ld-linux-x86-64.so.2 (0x00007f0240b02000)
```

No rpaths telling us where to find 
```
>> readelf -d /home/marco/.cache/cargo/debug/deps/test_proc-36289756094210ee

Dynamic section at offset 0x2628a0 contains 34 entries:
  Tag        Type                         Name/Value
 0x0000000000000001 (NEEDED)             Shared library: [libtest-c0479f7cd36f68a3.so]
 0x0000000000000001 (NEEDED)             Shared library: [libproc_macro-520699affe05435a.so]
 0x0000000000000001 (NEEDED)             Shared library: [libsyntax-7c956b82cf4d08b4.so]
 0x0000000000000001 (NEEDED)             Shared library: [libsyntax_pos-aab3d379f3226375.so]
 0x0000000000000001 (NEEDED)             Shared library: [libstd-52f862a21e09568c.so]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
    ...
```

**Comment by redshirtrob** [comment]
Thanks for the detailed response. It does sound like an upstream limitation. I hesitate to call it a bug because this might be a limitation on proc_macro, which syn/quote both seem to depend on.

For example, I built a binary based on the associated library from the example project and it has the same dynamic libs, which is not exactly surprising.

**Comment by mfarrugi** [comment]
I would say the fact that it was fixed in https://github.com/djc/askama/pull/77 admits it's a bug.

I don't think there is a valid use-case for having a runtime proc-macro dependency, and this borders on being a rustc bug.

**Comment by mfarrugi** [comment]
Closing this as an upstream issue, feel free to reopen if you think there is something we should do about it.

**Comment by jtzhpf** [comment]
2410周报

 1、本周进展：

- 本周将 MirChecker 跑了起来，并将其仓库中的案例做了测试（部分），验证了其能力。但如果代码中存在显式调用 crate，会报错`can't find crate`，仓库中的案例也有一部分因此无法进行分析，同时对于论文中提及的能检测到第三方 crate 的 bug 能力尚未验证。

![Uploading 截屏2024-03-21 15.23.32.png…]()

2、下周计划： 
- 解决 can't find crate 的问题，运行于第三方 crate 上，复现 MirChecker 的结果
- 除了 interval abstract domain，MirChecker 还用到了 octagon, polyhedra, linear_equalities, ppl_polyhedra, ppl_linear_congruences, pkgrid_polyhedra_linear_congruences 等，我对此不了解，还需要学习一下这方面的知识
- 继续看 Fast Summary-based Whole-program Analysis to Identify Unsafe Memory Accesses in Rust
